Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift

```diff
@@ -204,13 +198,42 @@ final class HTTP2ClientRequestHandler: ChannelDuplexHandler {
             self.runFinalAction(.close, context: context)
 
         case .succeedRequest(let finalAction, let finalParts):
+            // We can force unwrap the request here, as we have just validated in the state machine,
+            // that the request object is still present.
             self.request!.succeedRequest(finalParts)
             self.request = nil
             self.runTimeoutAction(.clearIdleReadTimeoutTimer, context: context)
             self.runFinalAction(finalAction, context: context)
         }
     }
 
+    private func sendRequestHead(_ head: HTTPRequestHead, startBody: Bool, context: ChannelHandlerContext) {
+        if startBody {
+            context.writeAndFlush(self.wrapOutboundOut(.head(head)), promise: nil)
+
+            // The above write might trigger an error, which may lead to a call to `errorCaught`,
+            // which in turn, may fail the request and pop it from the handler. For this reason
+            // we must check if the request is still present here.
+            guard let request = self.request else { return }
+            request.requestHeadSent()
+            request.resumeRequestBodyStream()
```

Comment:
Is it possible that as a result of writing we still have a request but are _not_ in a state where it would make sense to call either of these? `requestHeadSent()` _seems_ like it should be fine since that's just what we did, it's the `resumeRequestBodyStream()` I'm curious about.

Reply:
Good question... I guess the big one here would be a change in writability...

Reply:
On second thought. Nope this should be fine. The first is very much just informal, the second one triggers an action... Since the first is informal, no worries! ☮️